File: prob1.py
#  Time Complexity : O(n)
# Space Complexity :O(1)
# Leet Code: Yes

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)
        output = list()
        for i in range(n):
            idx = abs(nums[i]) - 1
            if nums[idx] > 0:
                nums[idx] *= -1

        for i in range(n):
            if nums[i] > 0:
                output.append(i+1)
            # else:
            #     nums[i] *= -1

        return output

File: prob1.java
// Time Complexity : O (n) , n = length of nums array
// Space Complexity : O (1) no extra space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, faced minor issues while coding, was checking nums[i] > 0 instead of nums[index]

import java.util.*;
public class prob1 {

    public static List<Integer> disappearedNumbers(int[] nums){
        List<Integer> result = new ArrayList<>();
        if(nums == null || nums.length == 0) return new ArrayList<>();
        for(int i = 0; i < nums.length; i++){
            int index = Math.abs(nums[i]) - 1;
            if(nums[index] > 0){
                nums[index] *= -1;
            }
        }
        for(int i = 0; i < nums.length; i++){
            if(nums[i] > 0){
                result.add(i+1);
            }
        }
        return result;
 
    }
    public static void main(String[] args) {
        int[] nums = {4,-3,-2,-7,8,2,-3,-1};
        List<Integer> result = disappearedNumbers(nums);
        System.out.println(result);
    }
    
}
